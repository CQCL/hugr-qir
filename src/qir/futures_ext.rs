use anyhow::{bail, ensure, Result};
use hugr::{
    extension::prelude::bool_t,
    ops::ExtensionOp,
    types::CustomType,
    HugrView,
};
use hugr_llvm::{
    emit::{EmitFuncContext, EmitOpArgs},
    inkwell::types::BasicTypeEnum,
    types::TypingSession,
};
use tket2_hseries::extension::futures::{FutureOpDef, EXTENSION_ID, FUTURE_TYPE_NAME};

use super::QirCodegenExtension;

impl QirCodegenExtension {
    pub fn convert_future_type<'c>(
        &self,
        session: TypingSession<'c, '_>,
        custom_type: &CustomType,
    ) -> Result<BasicTypeEnum<'c>> {
        ensure!(
            custom_type.extension() == &EXTENSION_ID
                && custom_type.name() == FUTURE_TYPE_NAME.as_str()
                && custom_type.args() == [bool_t().into()]
        );
        Ok(session.iw_context().bool_type().into())
    }

    pub fn emit_futures_op<'c, H: HugrView>(
        &self,
        context: &mut EmitFuncContext<'c, '_, H>,
        args: EmitOpArgs<'c, '_, ExtensionOp, H>,
        op: FutureOpDef,
    ) -> Result<()> {
        match op {
            FutureOpDef::Read => args.outputs.finish(context.builder(), [args.inputs[0]]),
            FutureOpDef::Dup => {
                let input = args.inputs[0];
                args.outputs.finish(context.builder(), [input, input])
            }
            FutureOpDef::Free => args.outputs.finish(context.builder(), []),
            _ => bail!("Unknown op: {op:?}"),
        }
    }
}
