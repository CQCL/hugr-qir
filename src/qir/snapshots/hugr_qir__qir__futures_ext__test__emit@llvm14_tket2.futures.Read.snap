---
source: src/qir/futures_ext.rs
expression: mod_str
---
; ModuleID = 'test_context'
source_filename = "test_context"

define { i32, {}, {} } @_hl.main.1(i1 %0) {
alloca_block:
  br label %entry_block

entry_block:                                      ; preds = %alloca_block
  %1 = select i1 %0, { i32, {}, {} } { i32 1, {} poison, {} undef }, { i32, {}, {} } { i32 0, {} undef, {} poison }
  ret { i32, {}, {} } %1
}
